### DISCUSSION TITLE
[Custom Pipeline] Passing uniform arrays through code

### DISCUSSION THREAD

**Discussion by panthavma** [bnpr/Malt#323] [Q&A] [answered]
Hello !
I'm trying to pass values to my shader, stored as an array. I've found a way for now by setting the uniform array directly (`array[id]`), but I think this is pretty inefficient performance-wise. Is there a better way to do this easily ?

```c
// Shader code
uniform int myArray[128];
```
```py
# Custom Pipeline code
MaterialShader.uniforms['myArray[0]'].set_value(16)
MaterialShader.uniforms['myArray[1]'].set_value(64)
MaterialShader.uniforms['myArray[2]'].set_value(48)
# versus something like MaterialShader.uniforms['myArray'].set_value([16,64,48])
```

**Comment by pragma37** [comment, marked as answer]
I think there's no way to set array uniforms like that in OpenGL.

In those cases I would use a Uniform Buffer Object:
https://github.com/bnpr/Malt/blob/Development/Malt/Render/Common.py
https://github.com/bnpr/Malt/blob/a987506a8380ba655aae2d092b870de880684b2d/Malt/Pipeline.py#L310
https://github.com/bnpr/Malt/blob/a987506a8380ba655aae2d092b870de880684b2d/Malt/Pipeline.py#L381

**Reply by panthavma** [reply]
Thanks !
I managed to make it work, but I have a second question now when packing ints, why aren't they packed by default when using ctypes (as in, why do I need the *4 and why did you use ivec4 when faced with a similar problem) ?

Here's my code that works for passing an array for future reference:
```py
# Python side
dataBuffer = (ctypes.c_int32 * (4 * BUFFER_SIZE))()
dataBuffer[10 * 4] = 314
dataUBO = UBO()
dataUBO.load_data(dataBuffer)
dataUBO.bind(TestShader.uniform_blocks["BLOCKNAME"])
```
```c
// GLSL Side
layout(std140) uniform BLOCKNAME
{
	uniform int variableName[BUFFER_SIZE];
};

// variableName[10] == 314
```

**Reply by pragma37** [reply]
Because the std140 layout requires each element to have a 16 bytes alignment:
https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)#Memory_layout
